subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java

```diff
@@ -316,4 +343,40 @@ private static ImmutableMap<String, String> mergeMaps(
         }
         return builder.buildKeepingLast();
     }
+
+    private static class BuildPropertyScope implements GradlePropertiesListener.PropertyScope.Build {
```

Comment:
🤔  This reads like a data class and we use it as a key in a map, but it does not implement hashcode/equals.

I also think it's a good practice to indicate the expectation of instances to be usable as keys by declaring hashcode/equals on the interface, like here:
https://github.com/gradle/gradle/blob/3343e9f68f78eb9d93ed6b36ed2f8eed202e1ad9/subprojects/core/src/main/java/org/gradle/api/internal/project/ProjectInternal.java#L268-L283

Reply:
If that is the intention, we should add a unit test that validates that

Reply:
But honestly, data classes should be data classes. I think identity by equality is an optimization that brings in more complexity, and it should be shown that the complexity is worth it by showing that the optimization makes a difference. I don't believe the difference would be measurable and that's why I'd vote for data classes instead.

Having said that, if we add a Controller test that check the identity equality contract, it's fine by me

Reply:
> Having said that, if we add a Controller test that check the identity equality contract, it's fine by me

I'll add a test in the follow-up PR.